rts/motoko-rts-macros/src/lib.rs

```diff
@@ -0,0 +1,145 @@
+use proc_macro::TokenStream;
+use quote::quote;
+
+#[proc_macro_attribute]
+pub fn ic_fn(_attr: TokenStream, input: TokenStream) -> TokenStream {
```

Comment:
Would it be possible/worth it to check that the argument/result types are restricted to the handful of legal Wasm types?

Reply:
In macros we only have the syntactic information, for example we know the names of the argument types but nothing else about those types. However the compiler already checks that types are "FFI safe", for example if I try something like

```rust
pub struct MyStruct {
    x: usize,
    y: String,
}

#[no_mangle]
extern "C" fn my_fn() -> MyStruct {
    MyStruct {
        x: 10,
        y: "test".to_owned(),
    }
}
```

Here it's not clear how to return this struct using C ABI, so the compiler generates this:

```
warning: `extern` fn uses type `MyStruct`, which is not FFI-safe
 --> src/lib.rs:7:26
  |
7 | extern "C" fn my_fn() -> MyStruct {
  |                          ^^^^^^^^ not FFI-safe
  |
  = note: `#[warn(improper_ctypes_definitions)]` on by default
  = help: consider adding a `#[repr(C)]` or `#[repr(transparent)]` attribute to this struct
  = note: this struct has unspecified layout
```